axlearn/cloud/gcp/dataflow_inference_custom.py

```diff
@@ -0,0 +1,205 @@
+        '"
+
+To use GPUs for your job:
+$ axlearn gcp dataflow start \
+    --bundler_spec=dockerfile=Dockerfile \
+    --bundler_spec=repo=${DOCKER_REPO} \
+    --bundler_spec=image=${DOCKER_IMAGE} \
+    --bundler_spec=target=dataflow \
+    --bundler_spec=allow_dirty=True \
+    --dataflow_spec=runner=DataflowRunner \
+    --dataflow_spec=dataflow_service_options=\
+    "worker_accelerator=type:nvidia-l4;count:1;install-nvidia-driver" \
+    -- "'python3 -m axlearn.cloud.gcp.dataflow_inference_custom \
+        --module=text.gpt.c4_trainer \
+        --config=fuji-7B-single \
+        --trainer_dir='gs://.../checkpoints/step_xxx' \
+        '"
+
+"""
+
+
+import copy
+import logging
+import warnings
+from typing import Any, Dict, Optional, Sequence
+
+import apache_beam as beam
+import jax
+from absl import app, flags
+from absl.flags import argparse_flags
+from apache_beam.ml.inference.base import ModelHandler, PredictionResult, RunInference
+from apache_beam.options.pipeline_options import PipelineOptions
+
+import axlearn.common.input_fake as input_fake
+import axlearn.common.launch_trainer as trainer_utils
+from axlearn.common.inference import InferenceRunner, MethodRunner
+from axlearn.common.utils import NestedTensor
+
+warnings.filterwarnings("ignore")
+
+
+class CustomModelHandler(ModelHandler[Dict, PredictionResult, Any]):
+    """Defines how to load a model and run inference"""
+
+    def __init__(self, flag_dict):
+        self._flag_dict = flag_dict
+
+    def _flag_values_from_dict(self, flag_values: dict) -> flags.FlagValues:
+        # Avoid mutating global FLAGS.
+        fv = copy.deepcopy(flags.FLAGS)
+        for k, v in flag_values.items():
+            try:
+                fv.set_default(k, v)
+            except flags._exceptions.UnrecognizedFlagError:
+                # Ignore unrecognized flags from other modules
+                pass
+        fv.mark_as_parsed()
+        return fv
+
+    def load_model(self) -> MethodRunner:
+        """Loads a pre-trained model in the desired type (MethodRunner in this case).
+        Reference: https://github.com/apple/axlearn/blob/main/axlearn/common/inference.py#L54
+
+        Returns:
+          An instance of MethodRunner.
+        """
+        # construct absl FlagValues from dict
+        flag_values = self._flag_values_from_dict(self._flag_dict)
+
+        module_config = trainer_utils.get_trainer_config(flag_values=flag_values)
+
+        # get InferenceRunner Config from Trainer Config and instantiate InferenceRunner
+        inference_runner_cfg = InferenceRunner.config_from_trainer(module_config)
+        inference_runner_cfg.init_state_builder.set(dir=flag_values.trainer_dir)
+        inference_runner = InferenceRunner(cfg=inference_runner_cfg, parent=None)
+
+        # create Method Runner only once
+        method_runner = inference_runner.create_method_runner(
+            method="predict", prng_key=jax.random.PRNGKey(1)
```

Comment:
Should we have a way to configure the PRNGKey?

Reply:
How do we do that? In my previous testing, changing this might break things